cpp/src/arrow/compute/api_scalar.cc

```diff
@@ -560,8 +560,12 @@ StrptimeOptions::StrptimeOptions(std::string format, TimeUnit::type unit,
 StrptimeOptions::StrptimeOptions() : StrptimeOptions("", TimeUnit::MICRO, false) {}
 constexpr char StrptimeOptions::kTypeName[];
 
+StructFieldOptions::StructFieldOptions(FieldRef field_ref)
+    : FunctionOptions(internal::kStructFieldOptionsType), field_ref(field_ref) {}
 StructFieldOptions::StructFieldOptions(std::vector<int> indices)
-    : FunctionOptions(internal::kStructFieldOptionsType), indices(std::move(indices)) {}
+    : StructFieldOptions(FieldRef(FieldPath(std::move(indices)))) {}
```

Comment:
Should we set both FieldRef and indices here? (for backwards compatibility of the indices field in the options class)